lib/shopify_api/bulk/query.ex

```diff
@@ -0,0 +1,195 @@
+        {:error, @polling_timeout_message}
+
+      error ->
+        error
+    end
+  end
+
+  # Shopify returns a single newline which gets stripped and we are left with garbage,
+  # handle it nicely here.
+  def fetch(nil), do: {:ok, ""}
+
+  def fetch({:error, _} = error), do: error
+  def fetch({:ok, url}), do: fetch(url)
+
+  def fetch(url) when is_binary(url) do
+    url
+    |> HTTPoison.get()
+    |> case do
+      {:ok, %{body: jsonl}} -> {:ok, jsonl}
+      error -> error
+    end
+  end
+
+  def parse_response!(""), do: []
+  def parse_response!({:ok, jsonl}), do: parse_response!(jsonl)
+  def parse_response!({:error, msg}), do: raise(ShopifyAPI.Bulk.QueryError, msg)
+
+  def parse_response!(jsonl) when is_binary(jsonl) do
+    jsonl
+    |> String.split("\n", trim: true)
+    |> Enum.map(&ShopifyAPI.JSONSerializer.decode!/1)
+  end
+
+  @spec status(AuthToken.t()) :: {:ok, status_response()} | {:error, any()}
+  def status(%AuthToken{} = token) do
+    token
+    |> ShopifyAPI.graphql_request(@polling_query, 1)
+    |> case do
+      {:ok, %{response: %{"currentBulkOperation" => response}}} ->
+        {:ok, response}
+
+      {:error, _} = error ->
+        error
+    end
+  end
+
+  defp handle_errors(resp) do
+    errors = get_in(resp.response, ["bulkOperationRunQuery", "userErrors"])
+
+    case Enum.empty?(errors) do
+      true -> :ok
+      false -> {:error, fetch_first_error(errors)}
+    end
+  end
+
+  defp bulk_query_string(query) do
+    query_string = ShopifyAPI.JSONSerializer.encode!(query)
+
+    """
+    mutation {
+      bulkOperationRunQuery(
+        query: #{query_string}
+      ) {
+        bulkOperation {
+          id
+          status
+        }
+        userErrors {
+          field
+          message
+        }
+      }
+    }
+    """
+  end
+
+  defp fetch_first_error(errors) do
+    errors
+    |> List.first()
+    |> Map.get("message")
+  end
+
+  defp poll(token, bulk_query_id, polling_rate, max_poll, depth \\ 0)
+
+  defp poll(_token, bulk_query_id, _, max_poll, depth) when max_poll == depth,
+    do: {:error, :timeout, bulk_query_id}
+
+  defp poll(token, bulk_query_id, polling_rate, max_poll, depth) do
+    Process.sleep(polling_rate)
+
+    case status(token) do
+      {:ok, %{"status" => "COMPLETED", "url" => url} = _response} -> {:ok, url}
+      _ -> poll(token, bulk_query_id, polling_rate, max_poll, depth + 1)
+    end
+  end
+
+  defp maybe_cancel(false, _, _), do: nil
+  defp maybe_cancel(true, token, bid), do: token |> cancel(bid) |> poll_till_cancel(token)
+
+  defp poll_till_cancel(resp, token, max_poll \\ 300, depth \\ 0)
+
+  defp poll_till_cancel({:ok, %{"bulkOperation" => %{"status" => "CANCELED"}}}, _token, _, _),
+    do: true
+
+  defp poll_till_cancel({:ok, %{"status" => "CANCELED"}}, _token, _, _), do: true
+
+  defp poll_till_cancel(_, _token, max_poll, depth) when max_poll == depth,
+    do: {:error, :cancelation_timedout}
```

Comment:
We aren't doing anything with the return values for this function, and therefore not bubbling up a cancel timeout.

Reply:
Correct, do we want to differentiate between a cancel poll timeout and a general one?